sdks/go/pkg/beam/core/runtime/harness/statecache/statecache.go

```diff
@@ -0,0 +1,172 @@
+	fnpb "github.com/apache/beam/sdks/v2/go/pkg/beam/model/fnexecution_v1"
+)
+
+// SideInputCache stores a cache of reusable inputs for the purposes of
+// eliminating redundant calls to the runner during execution of ParDos
+// using side inputs.
+type SideInputCache struct {
+	cache       map[string]*exec.ReusableInput
+	idsToTokens map[string]string
+	validTokens []string
+	mu          sync.Mutex
+	capacity    int
+}
+
+// Init makes the cache map and the map of IDs to cache tokens for the
+// SideInputCache. Should only be called once.
+func (c *SideInputCache) Init(cap int) error {
+	if cap <= 0 {
+		return errors.Errorf("capacity must be a positive integer, got %v", cap)
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.cache = make(map[string]*exec.ReusableInput, cap)
+	c.idsToTokens = make(map[string]string)
+	return nil
+}
+
+// Completely clears the list of valid tokens. Should be called when
+// starting to handle a new request.
+func (c *SideInputCache) clearValidTokens() {
+	c.validTokens = nil
+}
+
+// SetValidTokens clears the list of valid tokens then sets new ones, also updating the mapping of
+// transform and side input IDs to cache tokens in the process. Should be called at the start of every
+// new ProcessBundleRequest. If the runner does not support caching, the passed cache token values
+// should be empty and all get/set requests will silently be no-ops.
+func (c *SideInputCache) SetValidTokens(cacheTokens ...fnpb.ProcessBundleRequest_CacheToken) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.clearValidTokens()
+	for _, tok := range cacheTokens {
+		// User State caching is currently not supported, so these tokens are ignored
+		if tok.GetUserState() != nil {
+			continue
+		} else {
+			s := tok.GetSideInput()
+			transformID := s.GetTransformId()
+			sideInputID := s.GetSideInputId()
+			token := string(tok.GetToken())
+			c.setValidToken(transformID, sideInputID, token)
+		}
+	}
+}
+
+// setValidToken adds a new valid token for a request into the SideInputCache struct
+// and maps the transform ID and side input ID pairing to the cache token.
+func (c *SideInputCache) setValidToken(transformID, sideInputID, token string) {
+	idKey := transformID + sideInputID
+	c.idsToTokens[idKey] = token
+	c.validTokens = append(c.validTokens, token)
+}
+
+func (c *SideInputCache) makeAndValidateToken(transformID, sideInputID string) (string, bool) {
+	idKey := transformID + sideInputID
+	// Check if it's a known token
+	tok, ok := c.idsToTokens[idKey]
+	if !ok {
+		return "", false
+	}
+	// Check if the known token is valid for this request
+	for _, t := range c.validTokens {
+		if t == tok {
+			return tok, true
+		}
+	}
+	return "", false
+}
+
+// QueryCache takes a transform ID and side input ID and checking if a corresponding side
+// input has been cached. A query having a bad token (e.g. one that doesn't make a known
+// token or one that makes a known but currently invalid token) is treated the same as a
```

Comment:
In short: the runner decides what side input tokens are valid/invalid for a given bundle. In this current setup the map of IDs to tokens is disjointed from this concept of validity, it just keeps up with tokens that have been seen before. So an invalid token in this case would be a token that has been seen before but isn't valid for the current bundle. Of course, I now think that extra level may be unnecessary.

Reply:
So this is the issue with the makeAndValidateToken call, is that as implemented, a newer bundle with different side inputs will invalidate cache calls for side inputs to other existing bundles. And this can flip back and forth on boundary conditions as a result. (A bundles needs set A, the next bundle needs set B, the third needs set A again, invalidating all the B cache requests...)

So this implies we should roughly keep the active set of tokens for the active bundles. Say a map[token]int. (or map[token]int8) which says how many bundles are making use of a given token. At the start of the bundles we add one to the tokens to the active set, and when the bundle ends, we remove them (subtracting from the active counts), and deleting when the count reaches 0. (this is important for iterating/eviction, and to keep the map memory smaller).

That way, the `isValid` call can check validity by seeing if `c.validTokens[token] > 0`, prioritizing deletions for tokens that are not valid as you intend, rather than simply prioritizing "the most recent bundle".